pkg/network/usm/istio.go

```diff
@@ -0,0 +1,221 @@
+		// missed by the process monitor stream
+		processSync := time.NewTicker(scanTerminatedProcessesInterval)
+
+		defer func() {
+			processSync.Stop()
+			// Cleaning up all active hooks
+			m.registry.Clear()
+			// Stopping the process monitor (if we're the last instance)
+			processMonitor.Stop()
+			// marking we're finished.
+			m.wg.Done()
+		}()
+
+		for {
+			select {
+			case <-m.done:
+				return
+			case <-processSync.C:
+				m.sync()
+				m.registry.Log()
+			}
+		}
+	}()
+
+	log.Debugf("Istio monitoring enabled")
+}
+
+func (m *istioMonitor) Stop() {
+	if m == nil {
+		return
+	}
+
+	close(m.done)
+	m.wg.Wait()
+}
+
+// sync state of istioMonitor with the current state of procFS
+// the purpose of this method is two-fold:
+// 1) register processes for which we missed exec events (targeted mostly at startup)
+// 2) unregister processes for which we missed exit events
+func (m *istioMonitor) sync() {
+	deletionCandidates := m.registry.GetRegisteredProcesses()
+
+	_ = util.WithAllProcs(m.procRoot, func(pid int) error {
+		if _, ok := deletionCandidates[uint32(pid)]; ok {
+			// We have previously hooked into this process and it remains active,
+			// so we remove it from the deletionCandidates list, and move on to the next PID
+			delete(deletionCandidates, uint32(pid))
+			return nil
+		}
+
+		// This is a new PID so we attempt to attach SSL probes to it
+		m.handleProcessExec(uint32(pid))
+		return nil
+	})
+
+	// At this point all entries from deletionCandidates are no longer alive, so
+	// we should dettach our SSL probes from them
+	for pid := range deletionCandidates {
+		m.handleProcessExit(pid)
+	}
+}
+
+func (m *istioMonitor) handleProcessExec(pid uint32) {
+	path, ok := m.isEnvoyProcess(pid)
+	if !ok {
+		return
+	}
+
+	m.registry.Register(
+		path,
+		pid,
+		m.registerCB,
+		m.unregisterCB,
+	)
+}
+
+func (m *istioMonitor) handleProcessExit(pid uint32) {
+	// We avoid filtering PIDs here because it's cheaper to simply do a registry lookup
+	// instead of fetching a process name in order to determine whether it is an
+	// envoy process or not (which at the very minimum involves syscalls)
+	m.registry.Unregister(pid)
+}
+
+// isEnvoyProcess determines whether or not a given PID represents an envoy process
+// Aside from a boolean we also return the path of the executable
+//
+// TODO:
+// refine process detection heuristic so we can remove the number of false
+// positives. A common case that is likely worth optimizing for is filtering
+// out "vanilla" envoy processes, and selecting only envoy processes that are
+// running inside istio containers. Based on a quick inspection I made, it
+// seems that we could also search for "istio" in the cmdline string in addition
+// to "envoy", since the command line arguments look more or less the following:
+//
+// /usr/local/bin/envoy -cetc/istio/proxy/envoy-rev.json ...
+func (m *istioMonitor) isEnvoyProcess(pid uint32) (string, bool) {
```

Comment:
having a function called `is...` which check if the argument fulfills some condition and on top of that returning another value, is not a good pattern and not making a readable code.

Reply:
any suggestions?

Reply:
`getEnvoyPath`